Wrappers/Python/cil/optimisation/algorithms/SIRT.py

```diff
@@ -201,9 +204,13 @@ def update(self):
                 self.x=self.constraint.proximal(self.x, tau=1)
 
     def update_objective(self):
-        r"""Returns the objective
+        r"""Returns the weighted least squares objective
 
-        .. math:: \frac{1}{2}\|A x - b\|^{2}
+        .. math:: \frac{1}{2}\|A x - b\|^{2}_{M}
 
         """
-        self.loss.append(0.5*self.r.squared_norm())
+        y = self.operator.direct(self.x)
+        y.subtract(self.data, out = y)
+        wy=self.M.multiply(y)
+        self.objective.append(0.5* y.dot(wy))
```

Comment:
This is adding an additional 2 copies of the data in to memory so we need to try to improve it.

`wy` is the same as the residuals `r` in `update` and this is still in memory. So instead I think we can square `r` (in pace) and then weight it with M (again  in place). As `r` is not reused by the algorithm it's ok using the memory for this. The first step of the algorithm recomputes the new residuals in to `r`.

Reply:
I tried this but got numerical errors (effectively from dividing and multiplying by M). I will try to debug this again.